precompile/contract.go

```diff
@@ -0,0 +1,136 @@
+// Note: because we only ever read from [functions] there no lock is required to make it thread-safe.
+type statefulPrecompileWithFunctionSelectors struct {
+	fallback  *statefulPrecompileFunction
+	functions map[string]*statefulPrecompileFunction
+}
+
+// newStatefulPrecompileWithFunctionSelectors generates new StatefulPrecompile using [functions] as the available functions and [fallback]
+// as an optional fallback if there is no input data. Note: the selector of [fallback] will be ignored, so it is required to be left empty.
+func newStatefulPrecompileWithFunctionSelectors(fallback *statefulPrecompileFunction, functions []*statefulPrecompileFunction) StatefulPrecompiledContract {
+	// Ensure that if a fallback is present, it does not have a mistakenly populated function selector.
+	if fallback != nil && len(fallback.selector) != 0 {
+		panic(fmt.Errorf("fallback function cannot specify non-zero length function selector"))
+	}
+
+	// Construct the contract and populate [functions].
+	contract := &statefulPrecompileWithFunctionSelectors{
+		fallback:  fallback,
+		functions: make(map[string]*statefulPrecompileFunction),
+	}
+	for _, function := range functions {
+		_, exists := contract.functions[string(function.selector)]
+		if exists {
+			panic(fmt.Errorf("cannot create stateful precompile with duplicated function selector: %q", function.selector))
+		}
+		contract.functions[string(function.selector)] = function
+	}
+
+	return contract
+}
+
+// Run selects the function using the 4 byte function selector at the start of the input and executes the underlying function on the
+// given arguments.
+func (s *statefulPrecompileWithFunctionSelectors) Run(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	// If there is no input data present, call the fallback function if present.
+	if len(input) == 0 && s.fallback != nil {
+		return s.fallback.execute(accessibleState, caller, addr, nil, suppliedGas, readOnly)
+	}
+
+	// Otherwise, an unexpected input size will result in an error.
+	if len(input) < selectorLen {
+		return nil, suppliedGas, fmt.Errorf("missing function selector to precompile - input length (%d)", len(input))
+	}
+
+	// Use the function selector to grab the correct function
+	selector := input[:selectorLen]
+	functionInput := input[selectorLen:]
+	function, ok := s.functions[string(selector)]
+	if !ok {
+		return nil, suppliedGas, fmt.Errorf("invalid function selector %#x", selector)
+	}
+
+	return function.execute(accessibleState, caller, addr, functionInput, suppliedGas, readOnly)
+}
+
+// RequiredGas returns the amount of gas consumed by the underlying function
+func (s *statefulPrecompileWithFunctionSelectors) RequiredGas(input []byte) uint64 {
+	// If there is no input data present, call the fallback function if present.
+	if len(input) == 0 && s.fallback != nil {
```

Comment:
Why wouldn't we call the `function.requiredGas` here?

Reply:
The fallback function of a contract is only invoked if the length is 0. If the input length is not zero, then we require that the input includes a function selector so that it can invoke the correct function's required gas call. If the length of the input is between 0 and 4 bytes (exclusive) then it is not valid and we simply return 0 and it will return an error during execution.

Reply:
Returning an error during execution is not a perfect system, we could return an error directly from RequiredGas in that case. Open to refactoring there, but it should end up with the same result.